### PR TITLE
Add a read_only settings that deactivate write endpoint methods.

### PR DESCRIPTION
**Pull request by Natim**
Refs Kinto/kinto#235

 - [x] Test that migrate in read_only mode display warnings for readonly backends.
 - [x] Test that all write resource endpoint are deactivated
 - [x] Test that the read_only public settings is set accordingly
 - [x] Test that the heartbeat continue to works in ``read_only`` mode.

### CODE REVIEW

File: cliquet/scripts/cliquet.py [hunk 1]
```diff
@@ -15,6 +17,12 @@ def deprecated_init(env):
 
 def init_schema(env):
     registry = env['registry']
+    settings = registry.settings
+
+    if settings['read_only']:
+        message = 'You cannot migrate the database with a read_only setup.'
+        warnings.warn(message, ReadOnlyError)
+        return
```
Comment (Natim) [inline cliquet/scripts/cliquet.py]: We may want to return an error code there.

File: cliquet/errors.py [hunk 2]
```diff
@@ -72,6 +72,10 @@
 """
 
 
+class ReadOnlyError(Warning):
```
Comment (tarekziade) [inline cliquet/errors.py]: Since you are using this as a warning, and you inherit from Warning, I would call it ReadOnlyWarning instead

File: cliquet/scripts/cliquet.py [hunk 3]
```diff
@@ -15,6 +17,12 @@ def deprecated_init(env):
 
 def init_schema(env):
     registry = env['registry']
+    settings = registry.settings
+
+    if settings.get('read_only', False):
+        message = 'You cannot migrate the database with a read_only setup.'
+        warnings.warn(message, ReadOnlyError)
```
Comment (tarekziade) [inline cliquet/scripts/cliquet.py]: Warnings should be used just to warn about something, and let us run python in a mode where a warning is turned into an exception.

Here, you are doing a warning and exiting no matter what. If this is the only use case for ReadOnlyError, you should turn it into an Exception and simply raise it. And remove the explicit sys.exit(1) code, because raising an error exits with 1.

File: cliquet/errors.py (hunk 2 above)
Comment (tarekziade) [inline cliquet/errors.py]: Per my comment below, I think you want a plain exception here.

**Comment by leplatrem** [comment]
I realize we had not the same idea in mind :]

Could you please detail what motivated you to do it in Cliquet rather that in Kinto ?

In my opinion: 

* schema migration should not be ran on a read-only instance, it sounds obvious (and `make serve` is not used in production anyway)
* the `kinto.readonly_mode` is just an alias for a set of settings to disable endpoints
* in case of read only, the storage heartbeat can be replaced from Kinto init by a simpler one
* the readonly can be provided in kinto's hello view (if ever necessary)
* with the chosen strategy here, the readonly feature is spread among cliquet and kinto (*because implicit creation in default bucket will also have to be disabled there*)

I'm sure you had a good reason I don't see here :)

File: cliquet/permission/postgresql/__init__.py [hunk 4]
```diff
@@ -55,15 +55,16 @@ class Permission(PermissionBase):
...
         super(Permission, self).__init__(*args, **kwargs)
         self.client = client
+        self.read_only = read_only
 
     def initialize_schema(self):
         # Create schema
         here = os.path.abspath(os.path.dirname(__file__))
         schema = open(os.path.join(here, 'schema.sql')).read()
-        with self.client.connect() as conn:
+        with self.client.connect(readonly=self.read_only) as conn:
```
Comment (leplatrem) [inline cliquet/permission/postgresql/__init__.py]: Why do you pass `readonly=True` if it is to execute a schema ?

File: cliquet_docs/api/utilities.rst [hunk 5]
```diff
@@ -15,6 +15,7 @@ The returned value is a JSON mapping containing:
 - ``documentation``: The URL to the service documentation. (this document!)
 - ``settings``: a mapping with the values of relevant public settings for clients
     - ``batch_max_requests``: Number of requests that can be made in a batch request.
+    - ``read_only``: Tells client if the service is in read_only mode.
```
Comment (leplatrem) [inline cliquet_docs/api/utilities.rst]: nit: `Tells client if` is not necessary here

File: cliquet/scripts/cliquet.py (hunk 3 above)
Comment (Natim) [inline cliquet/scripts/cliquet.py]: Yes you are totally right, also raising an exception add a traceback where I would like to just write down a message, maybe printing to stderr and exiting would be enough, what do you think?

File: CHANGELOG.rst [hunk 6]
```diff
@@ -17,6 +17,7 @@ This document describes changes between each past release.
   ``*_pool_recycle``, ``*_pool_timeout`` to control connections pool
   behaviour.
 - Follow redirections in batch subrequests (fixes #511)
+- Add a ``read_only`` setting to run the service in read-only mode. (#525)
```
Comment (leplatrem) [inline CHANGELOG.rst]: nit: `readonly` works well in one word too. Don't we to add some suffix, like `mode` or `resources` ?

File: cliquet/scripts/cliquet.py [hunk 7]
```diff
@@ -15,6 +16,12 @@ def deprecated_init(env):
 
 def init_schema(env):
     registry = env['registry']
+    settings = registry.settings
+
+    if settings.get('read_only', False):
+        message = 'You cannot migrate the database with a read_only setup.'
```
Comment (leplatrem) [inline cliquet/scripts/cliquet.py]: nit: I don't think we use `You` in the error messages.

File: cliquet/scripts/cliquet.py [hunk 8]
```diff
@@ -15,6 +16,12 @@ def deprecated_init(env):
...
+        print(message, file=sys.stderr)
+        sys.exit(1)
```
Comment (leplatrem) [inline cliquet/scripts/cliquet.py]: Side note:  the cache backend requires some write access, but we expect it to have been migrated with the two other backends

File: cliquet_docs/api/utilities.rst [hunk 9]
```diff
@@ -15,6 +15,7 @@ The returned value is a JSON mapping containing:
 - ``documentation``: The URL to the service documentation. (this document!)
 - ``settings``: a mapping with the values of relevant public settings for clients
     - ``batch_max_requests``: Number of requests that can be made in a batch request.
+    - ``read_only``: read_only mode service status.
```
Comment (leplatrem) [inline cliquet_docs/api/utilities.rst]: What about `Only request with read operations are allowed` ?

File: cliquet_docs/reference/configuration.rst [hunk 10]
```diff
@@ -73,6 +73,22 @@ following setting should be declared in the ``.ini`` file:
     cliquet.record_mushroom_patch_enabled = false
 
 
+Setting the service in read_only mode
+:::::::::::::::::::::::::::::::::::::
+
+It is also possible to deploy a cliquet service in read_only mode.
+
+Instead of listing all the endpoint, you can use the ``read_only`` settings::
```
Comment (leplatrem) [inline cliquet_docs/reference/configuration.rst]: of having to specify every resources endpoint and verbs, the `<read only>` setting allows to make sure the services only allows read operations.

File: cliquet_docs/reference/configuration.rst [hunk 11]
```diff
@@ -73,6 +73,22 @@ following setting should be declared in the ``.ini`` file:
...
+
+    cliquet.read_only = true
+
+This will disable all resources endpoints that are not accessed with one of
+``GET``, ``OPTIONS``, or ``HEAD`` method.
+
+This settings will also make sure the heartbeat does not fails to
+write into the database.
```
Comment (leplatrem) [inline cliquet_docs/reference/configuration.rst]: This sentence is too related to the code IMO. 

However it could be worth mentionning here the expected setup for backend. For exemple, that the cache differs from storage and permissions...

File: cliquet/resource/viewset.py [hunk 12]
```diff
@@ -151,6 +151,9 @@ def is_endpoint_enabled(self, endpoint_type, resource_name, method,
 
         Uses the settings to tell so.
         """
+        if settings.get('readonly') and method not in self.readonly_methods:
```
Comment (leplatrem) [inline cliquet/resource/viewset.py]: nit: use `asbool()` in case you case `readonly == "false"`

File: cliquet_docs/reference/configuration.rst (hunk 11 above)
Comment (Natim) [inline cliquet_docs/reference/configuration.rst]: Does this sentence fits better:

    The readonly mode also register a readonly heartbeat for the storage and permission backend.

File: cliquet/scripts/cliquet.py [hunk 13]
```diff
@@ -15,10 +17,18 @@ def deprecated_init(env):
...
+    readonly_backends = ('storage', 'permission')
+    readonly_mode = asbool(settings.get('readonly', False))
 
     for backend in ('cache', 'storage', 'permission'):
         if hasattr(registry, backend):
-            getattr(registry, backend).initialize_schema()
+            if readonly_mode and backend in readonly_backends:
+                message = ('Cannot migrate the %s backend while '
+                           'in readonly mode.' % backend)
+                warnings.warn(message, Warning)
```
Comment (tarekziade) [inline cliquet/scripts/cliquet.py]: Don't pass the base Warning class here, warn() uses a UserWarning by default. If you want a specific Warning class you need to pass a XXXWarning or create your own

File: Makefile [hunk 14]
```diff
@@ -35,6 +35,7 @@ build-requirements:
 	/tmp/cliquet/bin/pip freeze | tail -n +2 > requirements.txt
 
 tests-once: install-dev
+	@rm -fr .coverage
```
Comment (leplatrem) [inline Makefile]: :+1:

File: cliquet/resource/viewset.py [hunk 15]
```diff
@@ -151,9 +152,13 @@ def is_endpoint_enabled(self, endpoint_type, resource_name, method,
 
         Uses the settings to tell so.
         """
+        if (asbool(settings.get('readonly')) and
+                method not in self.readonly_methods):
```
Comment (leplatrem) [inline cliquet/resource/viewset.py]: nit: I would recommend an intermediary variable here. Readability over brevity :)

File: cliquet/tests/resource/test_viewset.py [hunk 16]
```diff
@@ -268,6 +268,42 @@ def test_is_endpoint_enabled_returns_true_if_disabled(self):
                                                  config)
         self.assertTrue(is_enabled)
 
+    def test_is_endpoint_enabled_returns_false_for_put_if_readonly(self):
+        viewset = ViewSet()
+        config = {
+            'readonly': True
+        }
```
Comment (leplatrem) [inline cliquet/tests/resource/test_viewset.py]: micronit: would also work on a single line, no?

File: cliquet/tests/test_permission.py [hunk 17]
```diff
@@ -117,7 +132,10 @@ def test_can_remove_a_principal_for_a_user(self):
     def test_can_remove_a_unexisting_principal_to_a_user(self):
         user_id = 'foo'
         principal = 'bar'
+        principal2 = 'foobar'
+        self.permission.add_user_principal(user_id, principal2)
         self.permission.remove_user_principal(user_id, principal)
+        self.permission.remove_user_principal(user_id, principal2)
```
Comment (leplatrem) [inline cliquet/tests/test_permission.py]: Why this change ?

File: cliquet/tests/test_permission.py [hunk 18]
```diff
@@ -82,12 +82,27 @@ def test_backend_error_is_raised_anywhere(self):
...
-        self.assertTrue(self.permission.ping(self.request))
+        ping = heartbeat(self.permission)
+        self.assertTrue(ping(self.request))
+
+    def test_ping_returns_false_if_unavailable_in_readonly_mode(self):
+        self.request.registry.settings['readonly'] = 'true'
+        ping = heartbeat(self.permission)
+        with mock.patch.object(self.permission, 'user_principals',
+                               side_effect=exceptions.BackendError(
+                                   "Segmentation fault.")):
```
Comment (leplatrem) [inline cliquet/tests/test_permission.py]: nit: This indentation is not very python-friendly

Look what happens if we leave that :D https://twitter.com/tarek_ziade/status/659466245322985472

File: cliquet_docs/reference/configuration.rst [hunk 19]
```diff
@@ -73,6 +73,26 @@ following setting should be declared in the ``.ini`` file:
     cliquet.record_mushroom_patch_enabled = false
 
 
+Setting the service in readonly mode
+::::::::::::::::::::::::::::::::::::
+
+It is also possible to deploy a cliquet service in readonly mode.
```
Comment (leplatrem) [inline cliquet_docs/reference/configuration.rst]: nit: I think we use `*Cliquet*` elsewhere

File: cliquet_docs/reference/configuration.rst [hunk 20]
```diff
@@ -73,6 +73,26 @@ following setting should be declared in the ``.ini`` file:
...
+
+Instead of listing all the endpoint, you can use the ``readonly`` settings::
+
+    cliquet.readonly = true
+
+This will disable all resources endpoints that are not accessed with one of
```
Comment (leplatrem) [inline cliquet_docs/reference/configuration.rst]: every resources (not sure)

File: cliquet_docs/reference/configuration.rst [hunk 21]
```diff
@@ -73,6 +73,26 @@ following setting should be declared in the ``.ini`` file:
...
+``GET``, ``OPTIONS``, or ``HEAD`` method.
```
Comment (leplatrem) [inline cliquet_docs/reference/configuration.rst]: method or HTTP verb?

File: cliquet_docs/reference/configuration.rst [hunk 22]
```diff
@@ -73,6 +73,26 @@ following setting should be declared in the ``.ini`` file:
...
+
+This settings will also activate readonly heartbeat checks for the
```
Comment (leplatrem) [inline cliquet_docs/reference/configuration.rst]: setting (singular)

File: cliquet_docs/reference/configuration.rst [hunk 23]
```diff
@@ -73,6 +73,26 @@ following setting should be declared in the ``.ini`` file:
...
+permission and the storage backend.
+
+In case you need to use the cache backend in readonly mode, it will
+still needs read-write capabilities. Do not configure it if you do not
```
Comment (leplatrem) [inline cliquet_docs/reference/configuration.rst]: still need (infinitive) read-write permissions/privileges?

File: cliquet_docs/reference/configuration.rst [hunk 24]
```diff
@@ -73,6 +73,26 @@ following setting should be declared in the ``.ini`` file:
...
+
+    cliquet.readonly = true
+
+This will disable all resources endpoints that are not accessed with one of
+``GET``, ``OPTIONS``, or ``HEAD`` method.
+
+This settings will also activate readonly heartbeat checks for the
+permission and the storage backend.
+
+In case you need to use the cache backend in readonly mode, it will
```
Comment (leplatrem) [inline cliquet_docs/reference/configuration.rst]: *In case you need to use* does not fit here, nor with the rest of the paragraph

File: cliquet_docs/api/utilities.rst [hunk 25]
```diff
@@ -15,6 +15,7 @@ The returned value is a JSON mapping containing:
 - ``documentation``: The URL to the service documentation. (this document!)
 - ``settings``: a mapping with the values of relevant public settings for clients
     - ``batch_max_requests``: Number of requests that can be made in a batch request.
+    - ``readonly``: Only request with read operations are allowed.
```
Comment (leplatrem) [inline cliquet_docs/api/utilities.rst]: requests (plural)

**Comment by leplatrem** [comment]
LGTM (only nits and english)

Please squash into a single commit though :)

File: cliquet/tests/test_permission.py (hunk 17 above)
Comment (Natim) [inline cliquet/tests/test_permission.py]: Because of the coverage with the memory backend :(
